11-using-vllm/model/model.py

```diff
@@ -0,0 +1,182 @@
+                        server_up = True
+                        break
+                except httpx.RequestError:
+                    time.sleep(1)  # Wait for 1 second before retrying
+
+            if not server_up:
+                raise RuntimeError(
+                    "Server failed to start within the maximum allowed time."
+                )
+        else:
+            try:
+                result = subprocess.run(
+                    ["nvidia-smi"], capture_output=True, text=True, check=True
+                )
+                print(result.stdout)
+            except subprocess.CalledProcessError as e:
+                print(f"Command failed with code {e.returncode}: {e.stderr}")
+
+            self.model_args = AsyncEngineArgs(model=model_repo_id, **self._vllm_config)
+            self.llm_engine = AsyncLLMEngine.from_engine_args(self.model_args)
+            self.tokenizer = AutoTokenizer.from_pretrained(model_repo_id)
+
+            try:
+                result = subprocess.run(
+                    ["nvidia-smi"], capture_output=True, text=True, check=True
+                )
+                print(result.stdout)
+            except subprocess.CalledProcessError as e:
+                print(f"Command failed with code {e.returncode}: {e.stderr}")
+
+    async def predict(self, model_input):
+        if self.openai_compatible:
+            # if the key metrics: true is present, let's return the vLLM /metrics endpoint
+            if model_input.get("metrics", False):
+                response = await self._client.get(f"{self.vllm_base_url}/metrics")
+                return response.text
+
+            # convenience for Baseten bridge
+            if "model" not in model_input and "model" in self._vllm_config:
+                print(
+                    f"model_input missing model due to Baseten bridge, using {self._vllm_config['model']}"
+                )
+                model_input["model"] = self._vllm_config["model"]
+
+            stream = model_input.get("stream", False)
+            if stream:
+
+                async def generator():
+                    async with self._client.stream(
+                        "POST",
+                        f"{self.vllm_base_url}/v1/chat/completions",
+                        json=model_input,
+                    ) as response:
+                        async for chunk in response.aiter_bytes():
+                            if chunk:
+                                yield chunk
+
+                return generator()
+            else:
+                response = await self._client.post(
+                    f"{self.vllm_base_url}/v1/chat/completions",
+                    json=model_input,
+                )
+                return response.json()
+        else:
+            prompt = model_input.pop("prompt")
+            stream = model_input.pop("stream", True)
+
+            sampling_params = SamplingParams(**model_input)
+            idx = str(uuid.uuid4().hex)
+            chat = [
```

Comment:
I think folks are increasingly expecting a messages input as a valid input in addition to prompt, even though originally it was meant for OpenAI. What do you think about adding support for it, similar to what we do [here](https://github.com/basetenlabs/truss-examples/blob/main/lora/model/model.py).

Reply:
done